src/zkp/pienc.rs

```diff
@@ -390,107 +405,99 @@ mod tests {
 
         // The lower edge case works (2^ELL))
         let lower_bound = BigNumber::one() << ELL;
-        let (proof, input) = build_random_proof(&mut rng, lower_bound.clone())?;
-        let mut transcript = Transcript::new(b"PiEncProof");
-        assert!(proof.verify(&input, &(), &mut transcript).is_ok());
-        let (proof, input) = build_random_proof(&mut rng, -lower_bound)?;
-        let mut transcript = Transcript::new(b"PiEncProof");
-        assert!(proof.verify(&input, &(), &mut transcript).is_ok());
+        with_random_proof(&mut rng2, lower_bound.clone(), test_code_is_ok)?;
+
+        with_random_proof(&mut rng2, -lower_bound, test_code_is_ok)?;
 
         // The higher edge case fails (2^ELL+EPSILON)
         let upper_bound = BigNumber::one() << (ELL + EPSILON);
-        let (proof, input) = build_random_proof(&mut rng, upper_bound.clone())?;
-        let mut transcript = Transcript::new(b"PiEncProof");
-        assert!(proof.verify(&input, &(), &mut transcript).is_err());
-        let (proof, input) = build_random_proof(&mut rng, -upper_bound)?;
-        let mut transcript = Transcript::new(b"PiEncProof");
-        assert!(proof.verify(&input, &(), &mut transcript).is_err());
+        with_random_proof(&mut rng2, upper_bound.clone(), test_code_is_err)?;
 
+        with_random_proof(&mut rng2, -upper_bound, test_code_is_err)?;
         Ok(())
     }
 
     #[test]
     fn every_proof_field_matters() {
         let rng = &mut init_testing();
         let plaintext = random_plusminus_by_size(rng, ELL);
+        // Clone rng to make rust lifetime's happy.
+        let rng2 = &mut rng.clone();
```

Comment:
So my comment didn't do a good job at explaining the issue. The closure, which we need now, captures the rng mutably. But we need a different rng to pass to `with_random_proof` (Can't pass the same rng as it is already being borrowed.

Hence we need two. Let me know if this is an issue?

Reply:
Oh I see. Potential issues:
- If we clone the RNG and it's not reseeded, we might get "matching" randomness in the two different places where we use them, which could give us some hard-to-track-down bugs
- On the other hand, the cloning means that we still only have one seed if we want to debug a test

To avoid the first option we could do something like creating `rng2` via the `StdRng::with_seed` method and using a seed generated by `rng`. Dunno if it's actually going to be a problem though.